libsignal-protocol/src/crypto.rs

```diff
@@ -1,23 +1,266 @@
-use rand::RngCore;
+use crate::{
+    buffer::Buffer,
+    errors::{InternalError, InternalErrorCode},
+};
 use std::{
     os::raw::{c_int, c_void},
     pin::Pin,
     slice,
 };
 use sys::{signal_buffer, signal_crypto_provider};
+pub enum SignalCipherType {
+    AesCtrNoPadding,
+    AesCbcPkcs5,
+}
 
+impl From<i32> for SignalCipherType {
+    fn from(v: i32) -> Self {
+        match v as u32 {
+            sys::SG_CIPHER_AES_CTR_NOPADDING => {
+                SignalCipherType::AesCtrNoPadding
+            },
+            sys::SG_CIPHER_AES_CBC_PKCS5 => SignalCipherType::AesCbcPkcs5,
+            _ => unimplemented!("Unimplemented Signal Cipher Type"),
```

Comment:
Having a panic in `From` feels like an antipattern, especially when we're trying to convert from an integer to an enum. Can we use `TryFrom` and a custom error type (e.g. `UnknownCipherType(i32)`) instead?

Reply:
Yeah, i thought about it for a second, but i forgot about the `TryFrom`.

By the way i think that `i32` value passed from the c from the protocol implementation, i will edit it anyway.

Reply:
> By the way i think that i32 value passed from the c from the protocol implementation,

I think `sys::SG_CIPHER_AES_CTR_NOPADDING` being a `u32` is an artifact of the way bindgen translates `#define`'s. It's annoying, but because the `FOO` in `#define FOO 42` doesn't technically have a type (it's just a string composed of digits), bindgen must default to thinking it's a `u32` when the C library is using it in the place of `int` (`i32`).

Reply:
yeah, that is a little bit confusing, since the the `cipher: c_int` and the bindgen seems to think it is `u32` instead of `i32`. also here since all of the values is `+ve` (0, 1, ...) it dose not matter if we cast the `i32` to the `u32`, isn't that ok ?